ntoskrnl/io/iomgr/deviface.c

```diff
@@ -26,192 +27,542 @@ C_ASSERT(sizeof(L"{01234567-89ab-cdef-0123-456789abcdef}") == GUID_STRING_BYTES
+        goto Quit;
+    }
+
+    /* Copy the contents of an existing symbolic link */
+    RtlCopyUnicodeString(&SymbolicLink, SymbolicLinkName);
+
+    /* Separate it into its constituents */
+    Status = IopSeparateSymbolicLink(&SymbolicLink,
+                                     NULL,
+                                     NULL,
+                                     NULL,
+                                     &ReferenceString,
+                                     &ReferenceStringPresent,
+                                     NULL);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to separate symbolic link %wZ, Status 0x%08lx\n", &SymbolicLink, Status);
+        goto Quit;
+    }
+
+    /* Did we got a ref string? */
+    if (ReferenceStringPresent)
+    {
+        SymbolicLink.MaximumLength = SymbolicLink.Length = (USHORT)((ULONG_PTR)(ReferenceString.Buffer) - (ULONG_PTR)SymbolicLink.Buffer);
     }
     else
     {
-        RtlInitUnicodeString(&ReferenceString, L"#");
+        ReferenceString.Buffer = &Buffer;
+        ReferenceString.MaximumLength = ReferenceString.Length = sizeof(WCHAR);
+    }
+
+    /* Fix prefix: '\??\' -> '##?#' */
+    SymbolicLink.Buffer[0] = L'#';
+    SymbolicLink.Buffer[1] = L'#';
+    SymbolicLink.Buffer[2] = L'?';
+    SymbolicLink.Buffer[3] = L'#';
+
+    /* Try to open or create device interface keys */
+    if (Create)
+    {
+        Status = IopCreateRegistryKeyEx(&DeviceKeyHandle,
+                                        ClassHandle,
+                                        &SymbolicLink,
+                                        DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
+                                        REG_OPTION_NON_VOLATILE,
+                                        &DeviceKeyDisposition);
+    }
+    else
+    {
+        Status = IopOpenRegistryKeyEx(&DeviceKeyHandle,
+                                      ClassHandle,
+                                      &SymbolicLink,
+                                      DesiredAccess | KEY_ENUMERATE_SUB_KEYS);
     }
 
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &SubKeyName,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               *GuidKeyRealP,
-                               NULL);
-    Status = ZwCreateKey(DeviceKeyRealP,
-                         DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                         &ObjectAttributes,
-                         0,
-                         NULL,
-                         REG_OPTION_NON_VOLATILE,
-                         NULL);
     if (!NT_SUCCESS(Status))
     {
-        DPRINT1("Failed to open %wZ%wZ\\%wZ Status %x\n", &BaseKeyU, &GuidString, &SubKeyName, Status);
-        goto cleanup;
+        DPRINT1("Failed to create or open %wZ, Status 0x%08lx\n", &SymbolicLink, Status);
+        goto Quit;
+    }
+
+    /* Replace first '\' by '#' in reference string */
```

Comment:
The ReferenceString doesn't contain any other '\' characters, besides only the first one?

Reply:
You mean does it have other `\` backslashes? No, it shouldn't. Usually, it has that only at the first place, and then the remaining part comes, in case ref string exists. Otherwise, if it does not, it consists from a single `\` anyway, which is replaced by `#` for registry purposes here.